### PR TITLE
Add an `Auto` option to live_compile

### PR DESCRIPTION
**Pull request by jvoisin**
As suggested by @zeroSteiner

### CODE REVIEW

File: lib/msf/core/post/linux/compile.rb [hunk 1]
```diff
@@ -36,6 +36,16 @@ def live_compile?
...
+    if datastore['COMPILER'] == 'Auto'
+      if has_gcc?
+        datastore['COMPILER'] = 'gcc'
+      elsif has_clang?
+        datastore['COMPILER'] = 'clang'
+      else
+        fail_with Module::Failure::BadConfig, "Unable to find gcc nor clang on the remote target."
+      end
+    end
+
```
Comment (bcoles) [inline lib/msf/core/post/linux/compile.rb:48]: Clobbering `datastore` values is generally frowned upon.

Can we do this with a `compiler` variable within this method instead?

I'm not aware of any existing scenario where the calling module (or anything else outside of `upload_and_compile`) will need to know what compiler was used.

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
It looks like the `#live_compile?` method needs to be updated to check if the COMPILER option is Auto or one of the explicit options.

Right now, I'm getting this exception and the compiler isn't selected.


```
msf6 exploit(linux/local/af_packet_chocobo_root_priv_esc) > run

[*] Started reverse TCP handler on 192.168.250.134:4444 
[!] AutoCheck is disabled, proceeding with exploitation
[-] Exploit aborted due to failure: bad-config: Auto is not installed. Set COMPILE False to upload a pre-compiled executable.
[*] Exploit completed, but no session was created.
msf6 exploit(linux/local/af_packet_chocobo_root_priv_esc) > 
```

File: lib/msf/core/post/linux/compile.rb [hunk 2]
```diff
@@ -24,6 +36,8 @@ def live_compile?
     elsif datastore['COMPILER'] == 'clang' && has_clang?
       vprint_good 'clang is installed'
       return true
+    elsif datastore['COMPILER'] == 'Auto' && get_compiler != ''
+      return true
```
Comment (smcintyre-r7) [inline lib/msf/core/post/linux/compile.rb:40]: Using the rails [`#present?`](https://apidock.com/rails/Object/present%3F) method will ensure that the value return from `#get_compiler` is a non empty string, it also accounts for nil.

```suggestion
    elsif datastore['COMPILER'] == 'Auto' && get_compiler.present?
      return true
```

File: lib/msf/core/post/linux/compile.rb [hunk 3]
```diff
@@ -36,7 +50,12 @@ def live_compile?
   def upload_and_compile(path, data, compiler_args='')
     write_file "#{path}.c", strip_comments(data)
 
-    compiler_cmd = "#{datastore['COMPILER']} -o '#{path}' '#{path}.c'"
+    compiler = datastore['COMPILER']
+    if datastore['COMPILER'] == 'Auto'
+      compiler = get_compiler
+    end
```
Comment (smcintyre-r7) [inline lib/msf/core/post/linux/compile.rb:56]: This is where we'd likely want to fail if `#get_compiler` can't return a compiler and the module ignored/didn't call `#live_compile?`.

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]

**Review by smcintyre-r7 — APPROVED** [review]

**Comment by cdelafuente-r7** [comment]
# Release Notes

This adds an `Auto` option to the `Msf::Post::Linux::Compile` library. This enables automatic selection of the compiler according to what is available locally on the target system.

**Pull request merged**